server/mqtt.go

```diff
@@ -2416,21 +2485,54 @@ func (as *mqttAccountSessionManager) processSubs(sess *mqttSession, c *client,
 			}
 		}
 
-		// Load retained messages.
+		// Find retained messages.
 		if fromSubProto {
-			if err := loadRMS([]byte(f.filter)); err != nil {
+			addRMSubjects := func(subject string) error {
+				sub := &subscription{
+					client:  c,
+					subject: []byte(subject),
+					sid:     []byte(subject),
+				}
+				if err := c.addShadowSubscriptions(acc, sub, false); err != nil {
+					return err
+				}
+
+				subs := []*subscription{sub}
+				if len(sub.shadow) > 0 {
+					subs = append(subs, sub.shadow...)
+				}
+				for _, sub := range subs {
+					subject := string(sub.subject)
+					// Best-effort loading the messages, logs on errors (to c.srv), loads
+					// once for subject.
+					as.addRetainedSubjectsForSubject(rmSubjects, subject)
+					for _, ss := range sub.shadow {
+						as.addRetainedSubjectsForSubject(rmSubjects, string(ss.subject))
```

Comment:
0/5 NS: if it makes sense to process shadow subs for retained messages, since those live in the mqtt topic space? (or is it NATS subject namespace?) check.